code/modules/mecha/mecha.dm

```diff
@@ -787,26 +805,82 @@
+
+	var/obj/item/mecha_parts/component/armor/ArmC = internal_components[MECH_ARMOR]
+
+	var/temp_deflect_chance = deflect_chance
+	var/temp_damage_minimum = damage_minimum
+
+	if(!ArmC)
+		temp_deflect_chance = src.deflect_chance + (defense_mode ? 25 : 0)
+		temp_damage_minimum = src.damage_minimum
+
+	else
+		temp_deflect_chance = round(ArmC.get_efficiency() * ArmC.deflect_chance + (defense_mode ? 25 : 0))
+		temp_damage_minimum = round(ArmC.get_efficiency() * ArmC.damage_minimum) + src.damage_minimum
+
+	if(prob(temp_deflect_chance))		//Does your attack get deflected outright.
+		src.occupant_message("<span class='notice'>\The [W] bounces off [src.name].</span>")
+		to_chat(user, "<span class='danger'>\The [W] bounces off [src.name].</span>")
 		src.log_append_to_last("Armor saved.")
-/*
-		for (var/mob/V in viewers(src))
-			if(V.client && !(V.blinded))
-				V.show_message("The [W] bounces off [src.name] armor.", 1)
-*/
+
+	else if(W.force < temp_damage_minimum)	//Is your attack too PATHETIC to do anything. 3 damage to a person shouldn't do anything to a mech.
+		src.occupant_message("<span class='notice'>\The [W] bounces off the armor.</span>")
+		src.visible_message("\The [W] bounces off \the [src] armor")
+		return
+
 	else
-		src.occupant_message("<span class='red'><b>[user] hits [src] with [W].</b></span>")
-		user.visible_message("<span class='red'><b>[user] hits [src] with [W].</b></span>", "<span class='red'><b>You hit [src] with [W].</b></span>")
-		src.take_damage(W.force, damage_type = W.damtype)
-		src.check_for_internal_damage(list(MECHA_INT_TEMP_CONTROL,MECHA_INT_TANK_BREACH,MECHA_INT_CONTROL_LOST))
+		src.occupant_message("<font color='red'><b>[user] hits [src] with [W].</b></font>")
+		user.visible_message("<font color='red'><b>[user] hits [src] with [W].</b></font>", "<font color='red'><b>You hit [src] with [W].</b></font>")
+
+		var/pass_damage = W.force
+		for(var/obj/item/mecha_parts/mecha_equipment/armor/antiproj_armor_booster/ME in equipment)
+			pass_damage = ME.handle_projectile_contact(W, user, pass_damage)
+		src.take_damage(pass_damage,W.damtype)	//The take_damage() proc handles armor values
+		if(pass_damage > internal_damage_minimum)	//Only decently painful attacks trigger a chance of mech damage.
+			src.check_for_internal_damage(list(MECHA_INT_TEMP_CONTROL,MECHA_INT_TANK_BREACH,MECHA_INT_CONTROL_LOST))
 	return
 
-/obj/mecha/proc/get_remaining_equipment_slots()
-	if(equipment.len >= max_equip)
-		return 0
-	return max_equip - equipment.len
+//////////////////////////////////
+////////  Misc procs  ////////
+//////////////////////////////////
+
+/obj/mecha/proc/get_health()
+	return (health/initial(health)*100)
+
+/obj/mecha/proc/get_mecha_occupancy_state()
+	if((silicon_pilot) && silicon_icon_state)
+		return silicon_icon_state
+	if(occupant)
+		return icon_state
+	return "[icon_state]-open"
+
+/obj/mecha/proc/TryMaints(var/mob/user, var/obj/item/weapon/card/id/id_card)
+	if(!user in range(1))
```

Comment:
```suggestion
	if(!(user in range(1)))
```

Reply:
Aaaah that's how youre supposed to do tgat, thx